NUTCH-1465 Support sitemaps in Nutch

lewismc opened the pull request:
Hi Folks this issue addresses [NUTCH-1465](https://issues.apache.org/jira/browse/NUTCH-1465), I have an issue with some code which I will point out separately.

lewismc commented on src/java/org/apache/nutch/util/SitemapProcessor.java:
```diff
@@ -0,0 +1,436 @@
+            return;
+          }
+
+          BaseRobotRules rules = protocolFactory.getProtocol(url).getRobotRules(new Text(url), datum, new LinkedList<>());
```
Always passing a new LinkedList as the third parameter to the [.getRobotsRules](https://builds.apache.org/job/nutch-trunk/javadoc/org/apache/nutch/protocol/Protocol.html#getRobotRules-org.apache.hadoop.io.Text-org.apache.nutch.crawl.CrawlDatum-java.util.List-) method call may not be preferable. I've looked at the code and we have the option to pass null. This needs to be tested.
I have seen elsewhere in the codebase that use of this signature aligns with use of fetcher.store.robotstxt configuration property... so we may wish to so the same here and align it.

sebastian-nagel replied:
It's safe to pass null unless you want to use the robots.txt content.

lewismc commented on src/java/org/apache/nutch/util/SitemapProcessor.java:
```diff
@@ -0,0 +1,436 @@
+  public static final Logger LOG = LoggerFactory.getLogger(SitemapProcessor.class);
+  public static final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+
+  public static final String CURRENT_NAME = "current";
```
I also introduced this constant to mimic what is done in CrawlDb and LinkDb classes. This is means that represent the current HostDb... of course we don't have a HostDb class in the codebase right now so this constant has been introduced.

sebastian-nagel replied:
But in [ReadHostDb](https://github.com/apache/nutch/blob/master/src/java/org/apache/nutch/hostdb/ReadHostDb.java#L182) and [UpdateHostDb](https://github.com/apache/nutch/blob/master/src/java/org/apache/nutch/hostdb/UpdateHostDb.java#L107) still a String literal `"current"` is used.

lewismc replied:
What is your suggestion here @sebastian-nagel ?

lewismc commented:
We could also improve with parameterized logging in due course. I wanted to post this patch as a mechanism for relighting the interest in Sitemap parsing with master branch.

sebastian-nagel commented on src/java/org/apache/nutch/util/SitemapProcessor.java:
```diff
@@ -0,0 +1,436 @@
+        // the information from sitemap to the original datum. Emit the original crawl datum
+        if(sitemapDatum != null) {
+          context.getCounter("Sitemap", "existing_sitemap_entries").increment(1);
+          originalDatum.setScore(sitemapDatum.getScore());
```
- the [sitemap spec](https://www.sitemaps.org/protocol.html#xmlTagDefinitions) defines "the priority of this URL relative to other URLs on your site." That's different from a global score as calculated by OPIC, page rank, etc.
- overwriting fetchInterval will make [calculateLastFetchTime()](https://github.com/apache/nutch/blob/master/src/java/org/apache/nutch/crawl/AbstractFetchSchedule.java#L158) return the wrong time when a page has been fetched
- overwriting the modified time breaks any if-modified-since handling.

See the discussion in [NUTCH-1465](https://issues.apache.org/jira/browse/NUTCH-1465) on Jan 30-31, 2014. I was also wrong how to map these "concepts" from sitemaps to Nutch-internal ones.

lewismc commented:
@sebastian-nagel I've addressed all but two of your comments and responded. I've also implemented parameterized logging. In addition, I've dropped the STATUS_SITEMAP replacing instances with STATUS_INJECTED.
N.B. when I run this as follows i am not currently able to inject any URLs into the CrawlDB
```
//First I inject a random URL to create a CrawlDB

lmcgibbn@LMC-056430 /usr/local/nutch(NUTCH-1465) $ ./runtime/local/bin/nutch inject crawl urls/
Injector: starting at 2017-05-18 23:01:14
Injector: crawlDb: crawl
Injector: urlDir: urls
Injector: Converting injected urls to crawl db entries.
Injector: overwrite: false
Injector: update: false
Injector: Total urls rejected by filters: 0
Injector: Total urls injected after normalization and filtering: 1
Injector: Total urls injected but already in CrawlDb: 0
Injector: Total new urls injected: 1
Injector: finished at 2017-05-18 23:01:15, elapsed: 00:00:01

// I then, attempt to process a sitemap at http://www.autotrader.com/sitemap.xml which I've added to a file located in a 'sitemaps' directory

lmcgibbn@LMC-056430 /usr/local/nutch(NUTCH-1465) $ ./runtime/local/bin/nutch sitemap crawl -sitemapUrls sitemaps
SitemapProcessor: sitemap urls dir: sitemaps
SitemapProcessor: Starting at 2017-05-18 23:06:38
robots.txt whitelist not configured.
SitemapProcessor: Total records rejected by filters: 0
SitemapProcessor: Total sitemaps from HostDb: 0
SitemapProcessor: Total sitemaps from seed urls: 1
SitemapProcessor: Total failed sitemap fetches: 0
SitemapProcessor: Total new sitemap entries added: 0
SitemapProcessor: Finished at 2017-05-18 23:06:48, elapsed: 00:00:10

// Lets read the DB

lmcgibbn@LMC-056430 /usr/local/nutch(NUTCH-1465) $ ./runtime/local/bin/nutch readdb crawl -stats
CrawlDb statistics start: crawl
Statistics for CrawlDb: crawl
TOTAL urls:	1
shortest fetch interval:	30 days, 00:00:00
avg fetch interval:	30 days, 00:00:00
longest fetch interval:	30 days, 00:00:00
earliest fetch time:	Thu May 18 23:01:00 PDT 2017
avg of fetch times:	Thu May 18 23:01:00 PDT 2017
latest fetch time:	Thu May 18 23:01:00 PDT 2017
retry 0:	1
min score:	1.0
avg score:	1.0
max score:	1.0
status 1 (db_unfetched):	1
CrawlDb statistics: done
```
As you can see no URLs seem to be processed as the new sitemap entries added is zero, this is confirmed by the readdb output.
I need to do some more debugging and see where the bug(s) are. If anyone is able to try this patch out and has an interest in Sitemap support in Nutch master it would be highly appreciated.

marconett commented:
I'm running into the same problem and am unable to inject sitemap content into the db. here's the commands i used (not including output, it's the same as above):

```
bin/nutch inject crawl/crawldb urls/
bin/nutch sitemap crawl/crawldb -sitemapUrls sitemaps/ -noStrict -noFilter -noNormalize
bin/nutch readdb crawl/crawldb -stats
```

where `urls/seed.txt` contains "https://www.linux.com/" and `sitemaps/seed.txt` contains "https://www.linux.com/sitemap.xml".

I see (tcpdump) that there are https connections being established to linux.com while `bin/nutch sitemap` is running. But nothing gets injected into the crawldb.

Is there any info on this? Should this be fixed?

sebastian-nagel commented:
Hi @marconett, please ask for help on the [Nutch user mailing list](http://nutch.apache.org/mailing_lists.html) or report the problem at https://issues.apache.org/jira/projects/NUTCH. That's a closed pull request, and nothing will be fixed here. Thanks!
